### ISSUE TITLE
Allow multiple videos for the same video URL

### ISSUE THREAD

**Issue by bendk** [pculture/unisubs#2127]
We should allow multiple videos to be created for the same video URL.  This will allow the videos to exist in different teams and have separate subtitle data.

We can implement this by removing the unique_together constraint on VIdeoURL - url/type.  We need to replace it with a unique constraint on (team_video.team, video_url.url, video_url.type).  Since it's not really easy to implement this constraint directly in the database, especially since we're going through django's DB layer, so we should handle this in the python code.

Some additional issues:
 - This means that you can now have multiple public amara videos for a single video URL.  For example if there's one in the public area, and one on a public team.  This presents a question of which video to take subtitles from for the embedder.  Here's how we should handle this:
    - By default, we should prefer the public video
    - We should add an extra parameter when embedding videos, called data-team which allows one to specify which team to take the video from.  If that's present, then we should prefer a video from that team
    - Otherwise we should just pick an arbitrary video
- If a user moves a video to a team, we now need to check if there's another video for the same video URL.  If so we should raise an error and present the user with a message.  This applies to both the management forms and the API
- Similary, we should also check when a user removes a video from a team and puts it in the public area.

**Comment by franontanaya** [comment]
Currently adding some random token at the end of the URL does the trick (e.g. [?|&]amaratoken=something). Maybe that'll simplify things, and then later on this token can be used for something.

**Comment by kslottow** [comment]
This still seems relevant, as we want to discourage any workflows that move videos between teams.  Not something we'd want to focus on for v1.8 though (first release of Future UI pages to collab teams), is that correct?

**Comment by kslottow** [comment]
as this is related to https://github.com/pculture/amara-enterprise/issues/1115 I see that it should be done sooner rather than later.

**Comment by kslottow** [comment]
@bendk how would you estimate this one?

**Comment by kslottow** [comment]
@syl22-00 is this one you could do?  If so, how would you estimate?

**Comment by bendk** [comment]
One question with this one is how would the embedder work with multiple amara videos for the same video URL?  If a user wanted to use the embedder with a video URL tied to videos in multiple teams and/or the amara public area, what would they see?

I think the only reasonable way we can handle this is just pick an arbitrary non-private video and use those subtitles.  Is that okay?

One alternative that has been discussed on only allowing 1 public video and requiring extra copies of the video to always be in private teams.  But then how do we enforce that when team admins can change their teams from public to private?  We would have to check all of there videos and see if there was a duplicate, which seems really complicated both code wise and for the team admin.

**Comment by deanjansen** [comment]
@Ben, I'm not sure if this would solve the issue, but would it make sense to add a parameter for Amara ID (for the embedder), so a person could specify a specific instance of a video on Amara? And if they left it blank, maybe we default to the version of a video that was submitted earliest?

**Comment by kslottow** [comment]
@PCF-Testing what do you think about Ben's suggestion above for how to make the embedder work with a video URL tied to videos in multiple teams/Amara public area?

**Comment by bendk** [comment]
One last issue is now there is a possible error when moving videos to a different team, or removing them and putting them in the public area.  In either case, it's possible this would result in a duplicate video URL and therefore we should prevent it and return an error message.

**Comment by bendk** [comment]
Going to update the description with all this info.

**Comment by bendk** [comment]
@PCF-Testing Just pushed the start of code for this one.  It's ready for testing if you have slack time.

The code is gh-2127 on both repos.  There's a migration there.  It doesn't change the unique constraint yet, all it does is try to make sure we set the team attribute correctly on VideoUrl.  Here's how to test:

 - Run the migration
 - Run `dev manage set_videourl_team`.  This should set the VideoUrl.team column
 - You can now go to any futureui-style video page.  If any of the video url's have the wrong team set, and you're logged in as a staff user, then you should see a warning message about it.
 - The basic testing process should be to do something, then check that the team gets updated by visiting the futureui-style video page (AKA the video page with ?team in the URL -- it doesn't matter if it's set to a valid team or not)
  - Some things to test are: adding a video to a team, moving a video from team to team, removing a video from a team.

**Comment by bendk** [comment]
I'm guessing you haven't started testing, but I just pushed a couple changes.  I also changed the branch 
name to `gh-2127-prep`

**Comment by bendk** [comment]
Just pushed the rest of the code to the `gh-2127` branch

That branch actually updates the behavior to allow a URL to be in multiple teams.  The things to test are:
- Adding videos to the public area
  - If the video is already in a team the form should succeed as normal.  The
    new video shouldn't share subtitles with the existing team video
  - If the video is already in the public area, then the form should succeed
    and you should be redirected to the existing video page
 - Adding videos to teams
    - Adding Videos already in the public area, or from other teams should
      succeed.  We should create a new video that doesn't share subtitles with
      the other video
    - Adding videos already in your team:
        - Single URL form/API: fail with an error message
          saying the video was already added to your team
        - Multiple URL form/CSV import: note that the URL was already added to your team
- Move videos (management form and old-style move videos page)
  - If the video is already on the destination team, then we should display an error saying the video was aready added
- Remove videos (old, new remove videos forms, API)
  - If the video is already in the public area, then we should display an error saying the video was already added to the public area
- Moving public videos to your team
  - This is the form that we link to from old-style video pages.  Should even support this anymore now that you can just add the URL to your team and have a new video?
  - Assuming we support it, this one can now fail if the URL was already added to your team and now you try to move it from the public area.  In this case the form should fail and we should display an error message.

**Comment by bendk** [comment]
I also realized I needed to update the gh-2127-prep branch.  If you started testing it already, then please start over with the latest version.

This is one of those changes that we should do in two passes, so there are two distinct testing phases:

  - test gh-2127-prep -- check that we're setting the team_id column properly
 - test gh-2127 prep -- check the above stuff

**Comment by PCF-Testing** [comment]
On gh-2127-prep branch, videos added to old-style teams, have Team IDs for their URLs set to 0 (as for the public / community videos). Was this done intentionally, or is just an oversight?

**Comment by bendk** [comment]
Good catch, it should be fixed now.

**Comment by PCF-Testing** [comment]
The fix works for the new videos being added to the team, but not for the older videos that had been submitted before. Could you please also check manage set_videourl_team command, please?

Also, on new video pages after opening them for the first time after running set_videourl_team command, I see banner messages of "Updated team for http://www.youtube.com/watch?v=52a0Xjuaixo to Stakeholders Team" type. Is this as expected?

**Comment by bendk** [comment]
I fixed set_videourl_team to work properly.

The banner message should only show if there's a problem (and it also only shows to site admins).  So if you can find a reproducable way to see it after the latest changes, then report it and I'll try to fix it.

**Comment by PCF-Testing** [comment]
set_videourl_team works fine now. Error messages did not show up anymore the last time.

gh-2127-prep still has problems with a few unit tests (teams.tests.test_forms.TestEditVideosForm, teams.tests.test_forms.TestEditVideosForm, collab.tests.test_forms.EditEvaluationFormTest) that recently got fixed in 1409 and 1375.

**Comment by bendk** [comment]
I think that's fine.  It'll go away once we merge the code.

**Comment by PCF-Testing** [comment]
Branch gh-2127: when YouTube videos are added to old-style teams, subtitles get pulled from YT even if the hosting channel is not linked. This does not happen on staging or production.

**Comment by PCF-Testing** [comment]
gh-2127 generally looks good. Also checked that URL duplication does not happen within teams or the public area when the URLs are secondary for the video object.

One small issue I found is that submitting multiple URLs to collab team yields a wrong message - if of 2 submitted URLs one is new and one had existed i the team, the message says "1 videos successfully added, 0 videos already added to your team" (should be 1 and 1, respectively).

**Comment by kslottow** [comment]
@bendk just wanted to confirm which users this will be available to - will it be all teams (old-style and new), or only for collab teams using 1.8?

**Comment by bendk** [comment]
@kslottow it will be open to everyone.

@PCF-Testing I think I fixed the issue with the message when adding videos.   For the CSV import, do you see a message at the bottom of the email? That's where the error message should be.

**Comment by PCF-Testing** [comment]
Confirming the issue with adding multiple videos is fixed.

With import from CSV, the summary email to the admin correctly lists the videos that had existed on the team and were skipped during import, but still includes the skipped videos into the total number of imported videos:
 
--
The list of videos for team Stakeholders Team was processed completely. Here is a summary.Video is already added to team: http://www.youtube.com/watch?v=CkgCF64QLggNumber of videos added to team: 5
 
(5 videos in the above example is the total number of videos in the CSV, including the skipped one.)
This is a regression compared to production, although a small one.

The last and pretty important issue related to multiple videos per URL is interaction (import/export) with external systems (YouTube, Brightcove). On production, subtitles get fetched and pushed only for the YT channels that are linked to Amara, and only within the team that has the channel linked. On gh-2127, subtitles get fetched for some non-linked videos; also, if a channel is linked to a team and a copy of a video from the channel is added to the public area outside the team, subtitles are still fetched from YT. I think this should not happen, so that public copies of videos would not override "official" subtitles for the same videos created by enterprise teams. Syncing should behave in the same way as on production - for YT channels linked to user accounts, sync only subtitles from the public area; for channels linked from a team, sync only subtitles from specified teams.

**Comment by bendk** [comment]
@PCF-Testing can you explain the syncing issue some more, preferrably with instructions to reproduce.  Here's what I'm seeing on both staging and the branch, when a YT channel is linked to a team:

- If the video is added to the public area:
  - For public youtube videos we import the subtitles (maybe we want to change this behavior, I'm not sure, but it should be a separate ticket)
  - For private youtube videos, we don't import
  - When subtitles are added to the public video, we don't sync them

Are you seeing different results?  Is the issue with some other combination of things?

**Comment by bendk** [comment]
Also, I fixed the issue with the counts in the email after CSV import

**Comment by PCF-Testing** [comment]
OK, consider a collab team that has several of my YouTube channels linked to it (https://youtube.com/channel/UCDr5za7AQ0M2nYPphXmnjng, https://youtube.com/channel/UC2pr9bM8T-_BIWyfn2lyo7Q, https://youtube.com/channel/UCDSwJP7zmPBuojaZ2c2o6gg, https://youtube.com/channel/UCExhnOr29EvKAhKBiQAv2Dw).

When I add a YouTube video that is hosted by someone else's channel (e.g. https://www.youtube.com/watch?v=J7gJwbsXEFM or https://www.youtube.com/watch?v=h5gcJjOSDv4) to the above collab team, subtitles are pulled from YouTube with this video. In production, subtitles for such videos would not be pulled - only subtitles from linked channels are fetched, but not from any other channels.

When I add a video from my own linked channel, https://www.youtube.com/watch?v=tEajVRiaSaQ, to the above team, subtitles from that videos don't get fetched. This is a regression compared to production, where fetching works for collab teams. Syncing works on the branch.

When the same video from the linked channel is added to the public space, fetching does work (it wouldn't on production and I am not sure it should on the branch), syncing does not (which is OK IMO).

On old-style teams both fetching and syncing work well. On task teams, fetching is blocked by https://github.com/pculture/unisubs/issues/2722 (similarly to production).

So, the obvious problem is that on collab teams subtitles are being fetched where they shouldn't and are not fetched where they should.

**Comment by bendk** [comment]
For me, on staging, when I add a public youtube video, Amara always fetches the subtitles.  This happens both for team and non-team videos.  Have you tested this on staging?  Maybe there's a difference between that and production.

**Comment by bendk** [comment]
I'm also seeing subtitles getting fetched for teams when the account is linked on the gh-2127 branch.  Maybe I could make a demo branch and you could test it there.  That way if there was an issue then I could see what was happening easier.

**Comment by bendk** [comment]
Looks like there was two different issues with subtitle fetching.  As far as I can tell both should also be issues on staging.

- We fetch subtitles when we maybe shouldn't.  I made https://github.com/pculture/amara-enterprise/issues/1546 for this one.  
- We don't fetch subtitles when we should.  I just made 2aa731e to fix this.

**Comment by PCF-Testing** [comment]
Confirming that fetching subtitles from linked channels has been fixed.

Some videos do not load in the embedder on video and subtitle pages with the following error in the console:
![screenshot from 2017-07-27 07 46 42](https://user-images.githubusercontent.com/288455/28676778-6d1f0f18-72a0-11e7-99be-68b8c6ccef28.png)

Several unit tests are broken on the branch:

======================================================================
FAIL: test_forms (externalsites.tests.test_forms.AccountFormsetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/apps/amara/apps/externalsites/tests/test_forms.py", line 405, in test_forms
    'youtube_2',
AssertionError: Items in the first set but not the second:
'brightcovecms'

======================================================================
FAIL: test_team_account (externalsites.tests.test_models.GetSyncAccountTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/apps/amara/apps/externalsites/tests/test_models.py", line 46, in test_team_account
    self.check_get_sync_accounts(video, account)
  File "/opt/apps/amara/apps/externalsites/tests/test_models.py", line 39, in check_get_sync_accounts
    (account, video.get_primary_videourl_obj())
AssertionError: Lists differ: [] != [(<BrightcoveAccount: Brightco...

Second list contains 1 additional elements.
First extra element 0:
(<BrightcoveAccount: Brightcove: publisher>, <VideoUrl: http://bcove.me/services/link/bcpid1234/bctidabc>)

- []
+ [(<BrightcoveAccount: Brightcove: publisher>,
+   <VideoUrl: http://bcove.me/services/link/bcpid1234/bctidabc>)]

======================================================================
FAIL: test_user_account (externalsites.tests.test_models.GetSyncAccountTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/apps/amara/apps/externalsites/tests/test_models.py", line 52, in test_user_account
    self.check_get_sync_accounts(video, account)
  File "/opt/apps/amara/apps/externalsites/tests/test_models.py", line 39, in check_get_sync_accounts
    (account, video.get_primary_videourl_obj())
AssertionError: Lists differ: [] != [(<BrightcoveAccount: Brightco...

Second list contains 1 additional elements.
First extra element 0:
(<BrightcoveAccount: Brightcove: publisher>, <VideoUrl: http://bcove.me/services/link/bcpid1234/bctidabc>)

- []
+ [(<BrightcoveAccount: Brightcove: publisher>,
+   <VideoUrl: http://bcove.me/services/link/bcpid1234/bctidabc>)]

======================================================================
FAIL: test_fetch_subs (externalsites.tests.test_subfetch.SubFetchTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/apps/amara/apps/externalsites/tests/test_subfetch.py", line 81, in test_fetch_subs
    mock.call(self.account.oauth_refresh_token))
AssertionError: None != call('refresh-token')
-------------------- >> begin captured logging << --------------------
externalsites.subfetch: WARNING: fetch_subs() no available credentials.
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_fetch_subs_handles_bcp47_codes (externalsites.tests.test_subfetch.SubFetchTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/apps/amara/apps/externalsites/tests/test_subfetch.py", line 136, in test_fetch_subs_handles_bcp47_codes
    mock.call('test-access-token', 'caption-1'))
AssertionError: None != call('test-access-token', 'caption-1')
-------------------- >> begin captured logging << --------------------
externalsites.subfetch: WARNING: fetch_subs() no available credentials.
--------------------- >> end captured logging << ---------------------

**Comment by bendk** [comment]
Unittests fixed.  Can you give me steps to reproduce the embedder issue?

**Comment by PCF-Testing** [comment]
Some of the submitted videos do cause the problem with the embedder and others don't, without any obvious pattern.

The following videos do cause the issue (all submissions are made to a collab team with auto-creation of transcription request):
http://www.youtube.com/watch?v=-HparsK7m9I
https://www.youtube.com/watch?v=Vm8YtZlwcng
http://www.youtube.com/watch?v=ywpPxbo6FyA

The following videos don't:
http://www.youtube.com/watch?v=tprlQMClSYQ
http://www.youtube.com/watch?v=h5gcJjOSDv4
http://www.youtube.com/watch?v=FNp_JLft_tA
https://www.youtube.com/watch?v=5MgBikgcWnY
https://vimeo.com/225747214

**Comment by bendk** [comment]
It seems like it's working on the gh-2127 preview:

http://gh-2127.amara.org/en/videos/1CSiWfZNfz4I/info/table-tennis-timing-perfect-your-stroke/?team=m-collab

I'm not sure what the issue issue is, but I don't think it should block merging to staging.

**Comment by PCF-Testing** [comment]
Looks good except for https://github.com/pculture/amara-enterprise/issues/1539 which is not a regression.

**Comment by PCF-Testing** [comment]
Deployed.
